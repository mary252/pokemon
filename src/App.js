import React from 'react';
import axios from 'axios'
import { RingLoader } from 'react-spinners';
import './App.css';
import { thisExpression } from '@babel/types';

class App extends React.Component{
  state={
    poke_list:[],
    poke_data:[],
    current_pokemon_display:null,
    loading:true,
    chain:null,
    showEvolution:false,
    name:null,
    disabled:true
  }
  async componentDidMount(){
    this.setState({
      loading:true
    })
    axios.get(`${process.env.REACT_APP_URL}/pokemon/?limit=30`).then( res =>{
      //console.log(res)
      this.setState({
        poke_list:res.data.results
      })
    }).catch(err =>{
      console.log(err)
    })
    this.setState({
      loading:false
    })
  }
  fetchPokemonData= (index, url)=>{
    let{poke_data}=this.state
    let pokemon=poke_data.filter(poke =>poke.index==index)
    let acc=pokemon.length===0?
      axios.get(url).then( res =>{
        poke_data.push({
          index:index,
          name:res.data.forms[0].name,
          pic:res.data.sprites.front_default,
          states:res.data.stats,
          id:res.data.id
        })
      }).catch(err =>{
        console.log(err)
      }):null
  }
  draw_state= (states) =>{
   return states.map((stat,i)=>
    <div className="columns" key={i}>
      <div className="column is-6">{stat.stat.name}</div>
      <div className="column is-3">{stat.base_stat}</div>
    </div>
    )
  }
  FetchEvolution= async (id) =>{
    axios.get(`${process.env.REACT_APP_URL}/evolution-chain/${id}`).then(res =>{
      
      this.setState({
        chain:res.data.chain,
        disabled:false
      })
    })
  }
  showEvolution=()=>{
    let {chain}=this.state
    console.log(chain)
    return chain.evolves_to !=null && chain.evolves_to.length>0?
     <div className="">
      <p>{this.state.name}  </p>
      {this.continueEvolution(chain.evolves_to[0])}
    </div>:
    <div>no evolution</div>
  }
  continueEvolution=(chain)=>{
  return <p>{chain.species.name }
  {
    chain.evolves_to !=null && chain.evolves_to.length>0?
      this.continueEvolution(chain.evolves_to[0]):null
    
  }
  </p>
  }
  DisplayPokemon=( )=>{
    let pokemon=this.state.current_pokemon_display[0]
    //console.log(pokemon)
    this.FetchEvolution(pokemon.id)
    
    return <div className="states is-flex aic">
        <img className="pokemon-pic" src={pokemon.pic}/>
        <p className="pokemon-name">{pokemon.name}</p>
        {this.draw_state(pokemon.states)}
        <button className="evolution-button" onClick={()=>this.setState({
          showEvolution:true,
          name:pokemon.name
        })
        }
        disabled={this.state.disabled}>Show Evolution</button>
      </div>
  }
  draw_list=()=>{
   return this.state.poke_list.map((pokemon,i)=>
      <div 
        onMouseEnter={()=>this.fetchPokemonData(i,pokemon.url)} 
        onClick={()=> this.listClick(i)}
        className="list-entry columns"
        key={i}>
        <div className="column is-2">
          {i}
        </div>
        <div className="column is-5">
          {pokemon.name}
        </div>
      </div>
    )
  }
  listClick = (i) =>{
    let pokemon=this.state.poke_data.filter(poke =>poke.index==i)
    this.setState({
      current_pokemon_display:pokemon,
      showEvolution:false
    })
        
    this.props.history.push(`?pokemon=${pokemon[0].id}`);
  }
  render(){
    let {current_pokemon_display,loading,showEvolution}=this.state
 
    return (
      
        loading?
        <div className="pre-loader">
        <div className='sweet-loading'>
        <RingLoader
          color={'#ef5350'} 
          loading={this.state.loading} 
        />
      </div>
      </div>:
      
      <div className="container">
        
        <div className="header">
          <img className="header-pic" src={require("./assets/PokÃ©mon_logo.svg")}/>
        </div>
        <p className="label">Pokemons</p>
        <div className="box columns">
          <div className="column is-8">
            {this.draw_list()}
          </div>
          <div className="column is-4">
            {current_pokemon_display!=null?
             this.DisplayPokemon():null}
             <div className="is-flex aic">
             {
               showEvolution?this.showEvolution():null
             }
             </div>
          </div>
          
        </div>
      </div>
    );
  }
}

export default App;
