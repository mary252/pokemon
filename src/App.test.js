import React from 'react';
import {shallow } from 'enzyme';
import App from './App';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import axios from "axios";
import httpAdapter from "axios/lib/adapters/http";
import nock from "nock";
configure({ adapter: new Adapter() });
jest.setTimeout(30000);
const host = "https://pokeapi.co/api/v2";

axios.defaults.host = host;
axios.defaults.adapter = httpAdapter;
describe('renders without crashing', () => {
  
  const app = shallow(<App  />);

  let instance = app.instance();

  it("renders correctly",() => {
    expect(app).toMatchSnapshot();
  });
  it("component Did mount works",async()=>{
    nock("https://pokeapi.co/api/v2")
      .get("/pokemon/?limit=30")
      .reply(200, {
        data:{
          results: [
            {
              name:"bulbasaur",
              url:"https://pokeapi.co/api/v2/pokemon/1/"
            },
            {
              name:"ivysaur",
              url:"https://pokeapi.co/api/v2/pokemon/2/"
            }
          ]
        }
      });
    try{
      await instance.componentDidMount()
      expect(instance.state.loading).toEqual(false)
      expect(instance.state.poke_list.length).toEqual(2)
    }catch(e){
      
    }
    
  })

  it("fetch pokemon info work",async ()=>{
    nock("https://pokeapi.co/api/v2")
      .get("/pokemon/?limit=30")
      .reply(200, {
        data:{
          "base_experience": 64,
          "forms": [
            {
              "name": "bulbasaur",
              "url": "https://pokeapi.co/api/v2/pokemon-form/1/"
            }
          ],
          "id": 1,
          "sprites": {
            "front_default": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png",
          },
          "stats": [
            {
              "base_stat": 49,
              "effort": 0,
              "stat": {
                "name": "defense",
                "url": "https://pokeapi.co/api/v2/stat/3/"
              }
            },
            {
              "base_stat": 49,
              "effort": 0,
              "stat": {
                "name": "attack",
                "url": "https://pokeapi.co/api/v2/stat/2/"
              }
            },
            {
              "base_stat": 45,
              "effort": 0,
              "stat": {
                "name": "hp",
                "url": "https://pokeapi.co/api/v2/stat/1/"
              }
            }
          ],
        }
      });
      try{
        await instance.fetchPokemonData(0,"https://pokeapi.co/api/v2/pokemon/1/")
        expect(instance.state.poke_data.length).toEqual(1)
      }catch(e){

      }
  })

  it("fetch evolotion works",async ()=>{
    nock("https://pokeapi.co/api/v2")
      .get("/evolution-chain/1/")
      .reply(200, {
        data:{
          "chain": {
              "evolves_to": [
                {
                  "evolution_details": [
                    {
                      "gender": null,
                      "held_item": null,
                      "item": null,
                      "known_move": null,
                      "known_move_type": null,
                      "location": null,
                      "min_affection": null,
                      "min_beauty": null,
                      "min_happiness": null,
                      "min_level": 32,
                      "needs_overworld_rain": false,
                      "party_species": null,
                      "party_type": null,
                      "relative_physical_stats": null,
                      "time_of_day": "",
                      "trade_species": null,
                      "trigger": {
                        "name": "level-up",
                        "url": "https://pokeapi.co/api/v2/evolution-trigger/1/"
                      },
                      "turn_upside_down": false
                    }
                  ],
                  "evolves_to": [],
                  "is_baby": false,
                  "species": {
                    "name": "venusaur",
                    "url": "https://pokeapi.co/api/v2/pokemon-species/3/"
                  }
                }
              ]
        }
      }});
      try{
        await instance.FetchEvolution(1);
        expect(instance.state.chain).not.toEqual(null)
      }catch (e){

      }
  })
  afterEach(() => {
    delete global.__mobxInstanceCount; // prevent warnings
  })
});
