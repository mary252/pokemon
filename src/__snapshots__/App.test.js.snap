// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing renders correctly 1`] = `ShallowWrapper {}`;
